File: challenge/leet/easy/q697.py
'''
 * 697. Degree of an Array
 * https://leetcode.com/problems/degree-of-an-array
'''
import unittest


class Solution:
    def findShortestSubArray(sefl, nums: [int]) -> int:
        sz, fq = len(nums), 0;

        index = dict();

        for x, n in enumerate(nums):
            if n not in index:
                if fq == 0:
                    sz = 1
                index.update({n: (x, 1, 1)})
            else:
                freq = index[n][1] + 1
                dist = x - index[n][0] + 1
                index.update({n: (index[n][0], freq, dist)})
                if fq < freq:
                    fq = freq
                    sz = dist
                elif fq == freq:
                    sz = min(sz, dist)
        return sz


class MyTestCase(unittest.TestCase):
    def test_something(self):
        s = Solution()
        self.assertEqual(2, s.findShortestSubArray([1,2,2,3,1]))
        self.assertEqual(10, s.findShortestSubArray([1,2,2,3,1,4,2,1,2,1,2,1]))
        self.assertEqual(6, s.findShortestSubArray([1,2,2,3,1,4,2]))
        self.assertEqual(1, s.findShortestSubArray([2,1]))


if __name__ == '__main__':
    unittest.main()
    print("OK!")

File: challenge/leet.gcc/easy/q697/q697.cpp
/**
 * 697. Degree of an Array
 * https://leetcode.com/problems/degree-of-an-array
 */

#include <vector>
#include <cassert>
#include <map>
#include <iostream>

using namespace std;

struct Info
{
    int inf[3]; // index, frequence, length
};

class Solution {
public:
    int findShortestSubArray(vector<int>& nums) {
        map<int, Info> indx;
        int mn = nums.size(), fq = 0;

        for (int ix = 0; ix < nums.size(); ix++) {
            int n = nums[ix];
            if (indx.find(n) == indx.end()) {
                Info inf = { ix, 1, 1 };
                indx.emplace(pair<int, Info>(n, inf));

                if (fq < 1) {
                    mn = 1;
                    fq = 1;
                }
            }
            else {
                int dist = ix - indx[n].inf[0] + 1;
                indx[n].inf[2] = dist;

                indx[n].inf[1]++;
                int freq = indx[n].inf[1];

                if (fq < freq) {
                    mn = dist; // min(dist, mn);
                    fq = freq;
                }
                else if (fq == freq)
                    mn = min(dist, mn);
            }
        }
        return mn;
    }
};

int main() {
    Solution s;
    vector<int> nums;

    nums = vector<int>{ 2,1 };
    assert(1 == s.findShortestSubArray(nums));

    nums = vector<int>{ 1,2,2,3,1 };
    assert(2 == s.findShortestSubArray(nums));

    nums = vector<int>{ 1,2,2,3,2,7 };
    assert(4 == s.findShortestSubArray(nums));

    cout << "OK!" << endl;
    return 0;
}